### PR TITLE
 Avoid cycles while binding type parameter constraints

### PR DESCRIPTION
**Pull request by AlekseyTs**
This change consolidates two different techniques to avoid cycles into one. It is using special restricted mode of binding constraint types (proposed in https://github.com/dotnet/roslyn/pull/48270) but avoids the need to capture "partially" bound constraint types in symbols and the risk to use them when not appropriate. Instead, the information required to calculate IsValueType/IsReferenceType is immediately calculated from "partially" bound constraint types and preserved as special TypeParameterConstraintKind flags.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs [hunk 1]
```diff
@@ -285,13 +285,8 @@ internal bool IsEqualToOrDerivedFrom(TypeSymbol type, TypeCompareKind comparison
...
         /// You can ignore custom modifiers, ignore the distinction between object and dynamic, or ignore tuple element names differences.
         /// </param>
-        /// <param name="isValueTypeOverrideOpt">
-        /// A map from a type parameter symbol to a boolean value that should be used as a replacement for a value returned by
-        /// <see cref="TypeParameterSymbol.IsValueType"/> property. Used when accessing the property for a type parameter symbol
-        /// that has an entry in the map is not safe and can cause a cycle.  
-        /// </param>
         /// <returns>True if the types are equivalent.</returns>
-        internal virtual bool Equals(TypeSymbol t2, TypeCompareKind compareKind, IReadOnlyDictionary<TypeParameterSymbol, bool> isValueTypeOverrideOpt = null)
+        internal virtual bool Equals(TypeSymbol t2, TypeCompareKind compareKind)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs:289]: Thanks for catching that. It's a nice cleanup.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 3)

**Comment by AlekseyTs** [comment]
@cston Please take a look.

File: src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs [hunk 2]
```diff
@@ -310,37 +309,124 @@ internal static class ConstraintsHelper
...
+            Debug.Assert(!binder.Flags.Includes(BinderFlags.GenericConstraintsClause));
+            binder = binder.WithAdditionalFlags(BinderFlags.GenericConstraintsClause | BinderFlags.SuppressConstraintChecks);
+
+            ImmutableArray<TypeParameterConstraintClause> clauses;
+            clauses = binder.BindTypeParameterConstraintClauses(containingSymbol, typeParameters, typeParameterList, constraintClauses,
+                                                                diagnostics, performOnlyCycleSafeValidation: false);
+
+            if (clauses.All(clause => clause.ConstraintTypes.IsEmpty))
+            {
+                clauses = ImmutableArray<TypeParameterConstraintClause>.Empty;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs:336]: >clauses = ImmutableArray<TypeParameterConstraintClause>.Empty [](start = 16, length = 61)

`return ImmutableArray<TypeWithAnnotations>.Empty;`

File: src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs [hunk 3]
```diff
@@ -310,37 +309,124 @@ internal static class ConstraintsHelper
...
+                clauses = AdjustConstraintKindsBasedOnConstraintTypes(containingSymbol, typeParameters, clauses);
             }
 
-            // Wrap binder from factory in a generic constraints specific binder
-            // to avoid checking constraints when binding type names.
-            Debug.Assert(!binder.Flags.Includes(BinderFlags.GenericConstraintsClause));
-            binder = binder.WithAdditionalFlags(BinderFlags.GenericConstraintsClause | BinderFlags.SuppressConstraintChecks);
+            if (clauses.All(clause => clause.Constraints == TypeParameterConstraintKind.None))
+            {
+                clauses = ImmutableArray<TypeParameterConstraintClause>.Empty;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs:378]: >clauses = ImmutableArray<TypeParameterConstraintClause>.Empty; [](start = 16, length = 62)

`return ImmutableArray<TypeParameterConstraintKind>.Empty;`

File: src/Compilers/CSharp/Portable/Symbols/Source/TypeParameterConstraintClause.cs [hunk 4]
```diff
@@ -197,65 +183,74 @@ static bool isValueType(TypeParameterSymbol thisTypeParameter, ArrayBuilder<Type
...
-                    }, typePredicate: null, arg: (container, isValueTypeOverride), canDigThroughNullable: false, useDefaultType: true);
-                }
-            }
         }
-    }
 
-    internal static class TypeParameterConstraintClauseExtensions
-    {
-        internal static bool HasValue(this ImmutableArray<TypeParameterConstraintClause> constraintClauses, bool canIgnoreNullableContext)
+        internal static SmallDictionary<TypeParameterSymbol, bool> BuildIsReferenceTypeFromConstraintTypesMap(Symbol container, ImmutableArray<TypeParameterSymbol> typeParameters,
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/TypeParameterConstraintClause.cs:191]: >FromConstraintTypesMap [](start = 87, length = 22)

Is there a reason this method name contains "FromConstraintTypes" and the previous method name does not?

File: src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedTypeParameterSymbol.cs [hunk 5]
```diff
@@ -77,6 +77,14 @@ public override bool HasReferenceTypeConstraint
             }
         }
 
+        public override bool IsReferenceTypeFromConstraintTypes
+        {
+            get
+            {
+                return _underlyingTypeParameter.IsReferenceTypeFromConstraintTypes || CalculateIsReferenceTypeFromConstraintTypes(ConstraintTypesNoUseSiteDiagnostics);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedTypeParameterSymbol.cs:84]: >|| CalculateIsReferenceTypeFromConstraintTypes(ConstraintTypesNoUseSiteDiagnostics) [](start = 83, length = 83)

Is this necessary?

File: src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedTypeParameterSymbol.cs [hunk 6]
```diff
@@ -109,6 +117,14 @@ public override bool HasValueTypeConstraint
             }
         }
 
+        public override bool IsValueTypeFromConstraintTypes
+        {
+            get
+            {
+                return _underlyingTypeParameter.IsValueTypeFromConstraintTypes || CalculateIsValueTypeFromConstraintTypes(ConstraintTypesNoUseSiteDiagnostics);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedTypeParameterSymbol.cs:124]: >|| CalculateIsValueTypeFromConstraintTypes(ConstraintTypesNoUseSiteDiagnostics) [](start = 79, length = 79)

Is this necessary?

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EETypeParameterSymbol.cs [hunk 7]
```diff
@@ -59,6 +59,14 @@ public override bool HasReferenceTypeConstraint
             get { return _sourceTypeParameter.HasReferenceTypeConstraint; }
         }
 
+        public override bool IsReferenceTypeFromConstraintTypes
+        {
+            get
+            {
+                return _sourceTypeParameter.IsReferenceTypeFromConstraintTypes || CalculateIsReferenceTypeFromConstraintTypes(ConstraintTypesNoUseSiteDiagnostics);
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EETypeParameterSymbol.cs:66]: >|| CalculateIsReferenceTypeFromConstraintTypes(ConstraintTypesNoUseSiteDiagnostics) [](start = 79, length = 83)

Is this necessary?

File: src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedTypeParameterSymbol.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedTypeParameterSymbol.cs:84]: > Is this necessary? Yes, due to possible type substitution involved new types can appear in the constraint types

---
In reply to: [521607603](https://github.com/dotnet/roslyn/pull/49261#discussion_r521607603) [](ancestors = 521607603)

File: src/Compilers/CSharp/Portable/Symbols/Source/TypeParameterConstraintClause.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/TypeParameterConstraintClause.cs:191]: > Is there a reason this method name contains "FromConstraintTypes" and the previous method name does not?

Yes, this method only takes constraint types into account and the previous method does not.

---
In reply to: [521604502](https://github.com/dotnet/roslyn/pull/49261#discussion_r521604502) [](ancestors = 521604502)

File: src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedTypeParameterSymbol.cs (hunk 6 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedTypeParameterSymbol.cs:124]: > Is this necessary?

Yes, due to possible type substitution involved new types can appear in the constraint types

---
In reply to: [521607714](https://github.com/dotnet/roslyn/pull/49261#discussion_r521607714) [](ancestors = 521607714)

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EETypeParameterSymbol.cs (hunk 7 above)
Comment (AlekseyTs) [reply src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EETypeParameterSymbol.cs:66]: > Is this necessary?

Yes, due to possible type substitution involved new types can appear in the constraint types

---
In reply to: [521608683](https://github.com/dotnet/roslyn/pull/49261#discussion_r521608683) [](ancestors = 521608683)

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EETypeParameterSymbol.cs [hunk 8]
```diff
@@ -87,6 +95,14 @@ public override bool HasValueTypeConstraint
             get { return _sourceTypeParameter.HasValueTypeConstraint; }
         }
 
+        public override bool IsValueTypeFromConstraintTypes
+        {
+            get
+            {
+                return _sourceTypeParameter.IsValueTypeFromConstraintTypes || CalculateIsValueTypeFromConstraintTypes(ConstraintTypesNoUseSiteDiagnostics);
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EETypeParameterSymbol.cs:102]: > Is this necessary?

Yes, due to possible type substitution involved new types can appear in the constraint types

---
In reply to: [521608778](https://github.com/dotnet/roslyn/pull/49261#discussion_r521608778) [](ancestors = 521608778)

**Review by cston — APPROVED** [review]

**Pull request merged**